run.go

```diff
@@ -90,6 +89,17 @@ func (a *Autopilot) finishExecution(exec *execInfo) {
 	// of the world that we are now done
 	close(exec.done)
 	exec.done = nil
+
+	// We need to gain this lock so that we can nil out the previous state.
+	// This prevents us from accidentally tracking stale state in the event
+	// that we used to be the leader at some point in time, then weren't
+	// and now are again. In particular this will ensure that that we forget
+	// about our tracking of the firstStateTime so that once restarted, we
+	// will ignore server stabilization time just like we do the very
+	// first time this process ever was the leader.
+	a.stateLock.Lock()
+	defer a.stateLock.Unlock()
+	a.state = nil
```

Comment:
This `finishExecution` can be called when a different autopilot run tries to acquire the leaderLock but fails: https://github.com/hashicorp/raft-autopilot/blob/master/run.go#L100

That means it could be called by different autopilot runs and the one that fails to acquire the lock would empty out the state. 

Maybe this function should take a boolean to distinguish whether it is being called at the end of an execution or when the lock wasn't acquired?

Reply:
I think your right. In practice this shouldn't cause an issue but in theory it could so its best to change it up.

Reply:
I went ahead and moved the code into the deferred fn and out of finishExecution.

Reply:
The reason why we couldn't have hit this is because in order for a second beginExecution to be called from the Start method the other routine would have to be in a Stopping state. Therefore niling it an extra time during shutdown shouldn't matter and wont race because the state lock is held.

With all that being said, I think moving it to the deferred func that calls finishExecution is probably a little cleaner.